File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/Delphi/EditorsInPlaceValidationDemo/EditorsInPlaceValidationDemoMain.pas
unit EditorsInPlaceValidationDemoMain;

interface

uses
  Windows, Messages, SysUtils, Variants, Classes, Graphics, Forms, Controls,
  Dialogs, BaseForm, cxGraphics, cxControls, cxLookAndFeels, cxLookAndFeelPainters, cxStyles, cxCustomData, cxFilter,
  cxData, cxDataStorage, cxEdit, cxNavigator, cxGridCustomPopupMenu, cxGridPopupMenu, cxGridCustomTableView,
  cxGridTableView, cxGridCustomView, cxClasses, cxGridLevel, cxGrid, Menus, StdCtrls, cxTextEdit, cxGridCardView,
  ComCtrls, dxGDIPlusClasses, dxScreenTip, dxCustomHint, cxHint, ImgList, cxDropDownEdit;

type
  TEditorsInPlaceValidationDemoMainForm = class(TfmBaseForm)
    cxGrid: TcxGrid;
    cxGridLevel: TcxGridLevel;
    cxGridPopupMenu1: TcxGridPopupMenu;
    cxGridTableView: TcxGridTableView;
    cxGridTableViewColumnAddress: TcxGridColumn;
    cxGridTableViewColumnEmail: TcxGridColumn;
    cxGridTableViewColumnFirstName: TcxGridColumn;
    cxGridTableViewColumnLastName: TcxGridColumn;
    cxGridTableViewColumnPhoneNumber: TcxGridColumn;
    cxHintStyleController: TcxHintStyleController;
    dxScreenTipRepository: TdxScreenTipRepository;
    miOptions: TMenuItem;
    miValidationAllowLoseFocus: TMenuItem;
    miValidationRaiseException: TMenuItem;
    miValidationShowErrorIcon: TMenuItem;
    stGrid: TdxScreenTip;
    ValidationOptions1: TMenuItem;
    icCustomIconList: TcxImageCollection;
    icCustomIcon1: TcxImageCollectionItem;
    procedure cxGridTableViewColumnAddressPropertiesValidate(Sender: TObject; var DisplayValue: Variant;
      var ErrorText: TCaption; var Error: Boolean);
    procedure cxGridTableViewColumnAddressValidateDrawValue(Sender: TcxCustomGridTableItem;
      ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxGridTableViewColumnEmailPropertiesValidate(Sender: TObject; var DisplayValue: Variant;
      var ErrorText: TCaption; var Error: Boolean);
    procedure cxGridTableViewColumnEmailValidateDrawValue(Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord;
      const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxGridTableViewColumnFirstNamePropertiesValidate(Sender: TObject; var DisplayValue: Variant;
      var ErrorText: TCaption; var Error: Boolean);
    procedure cxGridTableViewColumnFirstNameValidateDrawValue(Sender: TcxCustomGridTableItem;
      ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxGridTableViewColumnLastNamePropertiesValidate(Sender: TObject; var DisplayValue: Variant;
      var ErrorText: TCaption; var Error: Boolean);
    procedure cxGridTableViewColumnLastNameValidateDrawValue(Sender: TcxCustomGridTableItem;
      ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxGridTableViewColumnPhoneNumberPropertiesValidate(Sender: TObject; var DisplayValue: Variant;
      var ErrorText: TCaption; var Error: Boolean);
    procedure cxGridTableViewColumnPhoneNumberValidateDrawValue(Sender: TcxCustomGridTableItem;
      ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
    procedure cxHintStyleControllerShowHintEx(Sender: TObject; var Caption, HintStr: string; var CanShow: Boolean;
      var HintInfo: THintInfo);
    procedure FormCreate(Sender: TObject);
  private
    function DoAddressValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoEmailValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoFirstNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoLastNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function DoPhoneNumberValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
    function GetPersonFullName(ARecord: TcxCustomGridRecord): string;
  published
    procedure InitializeEditors(Sender: TObject);
  end;

var
  EditorsInPlaceValidationDemoMainForm: TEditorsInPlaceValidationDemoMainForm;

implementation

{$R *.dfm}

uses
  cxRegExpr, cxGridRows;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnAddressPropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
begin
  Error := DoAddressValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnAddressValidateDrawValue(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoAddressValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetInfo;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnEmailPropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
begin
  Error := DoEmailValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnEmailValidateDrawValue(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoEmailValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetCustom;
    AData.ErrorIcon.Assign(icCustomIcon1.Picture.Bitmap);
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnFirstNamePropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
begin
  Error := DoFirstNameValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnFirstNameValidateDrawValue(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoFirstNameValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnLastNamePropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
begin
  Error := DoLastNameValidate(DisplayValue, ErrorText);
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnLastNameValidateDrawValue(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AErrorText: TCaption;
begin
  if DoLastNameValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetError;
    AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnPhoneNumberPropertiesValidate(Sender: TObject;
  var DisplayValue: Variant; var ErrorText: TCaption; var Error: Boolean);
var
  AFullName: string;
begin
  Error := DoPhoneNumberValidate(DisplayValue, ErrorText);
  AFullName := GetPersonFullName(cxGridTableView.ViewData.EditingRecord);
  if Trim(AFullName) > '' then
    ErrorText := ErrorText + ' for ' + AFullName;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxGridTableViewColumnPhoneNumberValidateDrawValue(
  Sender: TcxCustomGridTableItem; ARecord: TcxCustomGridRecord; const AValue: Variant; AData: TcxEditValidateInfo);
var
  AFullName: string;
  AErrorText: TCaption;
begin
  if DoPhoneNumberValidate(AValue, AErrorText) then
  begin
    AData.ErrorType := eetWarning;
    AFullName := GetPersonFullName(ARecord);
    if Trim(AFullName) > '' then
      AData.ErrorText := AErrorText + ' for ' + AFullName
    else
      AData.ErrorText := AErrorText;
  end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.cxHintStyleControllerShowHintEx(Sender: TObject; var Caption,
  HintStr: string; var CanShow: Boolean; var HintInfo: THintInfo);
begin
  stGrid.Header.Glyph.Assign(nil);
  if GetMenuItemChecked(miValidationShowErrorIcon) and (TObject(HintInfo.HintData) is TcxGridDataCellViewInfo) then
    case TcxGridDataCellViewInfo(HintInfo.HintData).EditViewInfo.ErrorData.ErrorType of
      eetError:
        stGrid.Header.Glyph.Assign(cxEditErrorIcon);
      eetWarning:
        stGrid.Header.Glyph.Assign(cxEditWarningIcon);
      eetInfo:
        stGrid.Header.Glyph.Assign(cxEditInfoIcon);
      eetCustom:
        stGrid.Header.Glyph.Assign(icCustomIcon1.Picture.Bitmap);
    end;
end;

procedure TEditorsInPlaceValidationDemoMainForm.FormCreate(Sender: TObject);
begin
  InitializeEditors(Self);
end;

procedure TEditorsInPlaceValidationDemoMainForm.InitializeEditors(Sender: TObject);
var
  AValidationOptions: TcxEditValidationOptions;
begin
  AValidationOptions := [];
  if GetMenuItemChecked(miValidationRaiseException) then
    Include(AValidationOptions, evoRaiseException);
  if GetMenuItemChecked(miValidationShowErrorIcon) then
    Include(AValidationOptions, evoShowErrorIcon);
  if GetMenuItemChecked(miValidationAllowLoseFocus) then
    Include(AValidationOptions, evoAllowLoseFocus);

  cxGridTableViewColumnFirstName.Properties.ValidationOptions := AValidationOptions;
  cxGridTableViewColumnLastName.Properties.ValidationOptions := AValidationOptions;
  cxGridTableViewColumnAddress.Properties.ValidationOptions := AValidationOptions;
  cxGridTableViewColumnPhoneNumber.Properties.ValidationOptions := AValidationOptions;
  cxGridTableViewColumnEmail.Properties.ValidationOptions := AValidationOptions;
end;

function TEditorsInPlaceValidationDemoMainForm.DoAddressValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
begin
  Result := (cxGridTableViewColumnAddress.Properties as TcxComboBoxProperties).Items.IndexOf(VarToStr(AValue)) = -1;
  if Result then
    AErrorText := 'Please select an address from the list';
end;

function TEditorsInPlaceValidationDemoMainForm.DoEmailValidate(const AValue: Variant;
  var AErrorText: TCaption): Boolean;
var
  S: string;
begin
  S := VarToStr(AValue);
  Result := (S <> '') and not IsTextFullValid(S, '[A-z0-9_-]+@[A-z0-9_-]+\.[A-z0-9_-]+(\.[A-z]+)*');
  if Result then
    AErrorText := 'Please enter a valid email address';
end;

function TEditorsInPlaceValidationDemoMainForm.DoFirstNameValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
begin
  Result := VarToStr(AValue) = '';
  if Result then
    AErrorText := 'Please enter a value';
end;

function TEditorsInPlaceValidationDemoMainForm.DoLastNameValidate(const AValue: Variant;
  var AErrorText: TCaption): Boolean;
begin
  Result := VarToStr(AValue) = '';
  if Result then
    AErrorText := 'Please enter a value';
end;

function TEditorsInPlaceValidationDemoMainForm.DoPhoneNumberValidate(const AValue: Variant; var AErrorText: TCaption): Boolean;
var
  S: string;
begin
  S := VarToStr(AValue);
  Result := (S = '') or not IsTextValid(S, '(\(\d\d\d\)'' '')?\d\d\d-\d\d\d\d');
  if Result then
    AErrorText := 'Please enter a valid phone number';
end;

function TEditorsInPlaceValidationDemoMainForm.GetPersonFullName(ARecord: TcxCustomGridRecord): string;
var
  AFirstName, ALastName: string;
begin
  AFirstName := VarToStr(ARecord.Values[cxGridTableViewColumnFirstName.Index]);
  ALastName := VarToStr(ARecord.Values[cxGridTableViewColumnLastName.Index]);
  if (Trim(AFirstName) > '') and (Trim(ALastName) > '') then
    Result := Format('%s %s', [AFirstName, ALastName])
  else
    Result := AFirstName + ALastName;
end;

end.

File: Components/DevExpressVCL/Demos/ExpressQuantumGrid/CBuilder/EditorsInPlaceValidationDemo/EditorsInPlaceValidationDemoMain.cpp
//---------------------------------------------------------------------------

#include <vcl.h>
#pragma hdrstop

#include "EditorsInPlaceValidationDemoMain.h"
//---------------------------------------------------------------------------
#pragma package(smart_init)
#pragma link "cxClasses"
#pragma link "cxGridCardView"
#pragma link "cxGridTableView"
#pragma link "cxStyles"
#pragma link "cxControls"
#pragma link "cxCustomData"
#pragma link "cxData"
#pragma link "cxDataStorage"
#pragma link "cxDropDownEdit"
#pragma link "cxEdit"
#pragma link "cxFilter"
#pragma link "cxGraphics"
#pragma link "cxGrid"
#pragma link "cxGridCustomPopupMenu"
#pragma link "cxGridCustomTableView"
#pragma link "cxGridCustomView"
#pragma link "cxGridLevel"
#pragma link "cxGridPopupMenu"
#pragma link "cxHint"
#pragma link "cxLookAndFeelPainters"
#pragma link "cxLookAndFeels"
#pragma link "cxNavigator"
#pragma link "cxTextEdit"
#pragma link "dxCustomHint"
#pragma link "dxScreenTip"
#pragma link "cxRegExpr"
#pragma link "cxGridRows"
#pragma resource "*.dfm"
TEditorsInPlaceValidationDemoMainForm *EditorsInPlaceValidationDemoMainForm;
//---------------------------------------------------------------------------
__fastcall TEditorsInPlaceValidationDemoMainForm::TEditorsInPlaceValidationDemoMainForm(TComponent* Owner)
	: TfmBaseForm(Owner)
{
}
//---------------------------------------------------------------------------
void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnAddressPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoAddressValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnAddressValidateDrawValue(TcxCustomGridTableItem *Sender,
          TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoAddressValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetInfo;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnEmailPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoEmailValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnEmailValidateDrawValue(TcxCustomGridTableItem *Sender,
          TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoEmailValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetCustom;
	AData->ErrorIcon->Assign(icCustomIcon1->Picture->Bitmap);
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnFirstNamePropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoFirstNameValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnFirstNameValidateDrawValue(TcxCustomGridTableItem *Sender,
          TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoFirstNameValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetError;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnLastNamePropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoLastNameValidate(DisplayValue, ErrorText);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnLastNameValidateDrawValue(TcxCustomGridTableItem *Sender,
          TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoLastNameValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetError;
	AData->ErrorText = AErrorText;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnPhoneNumberPropertiesValidate(TObject *Sender,
          Variant &DisplayValue, TCaption &ErrorText, bool &Error)
{
  Error = DoPhoneNumberValidate(DisplayValue, ErrorText);
  String AFullName = GetPersonFullName(cxGridTableView->ViewData->EditingRecord);
  if (Trim(AFullName) > "")
  {
	ErrorText = ErrorText + " for " + AFullName;
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxGridTableViewColumnPhoneNumberValidateDrawValue(TcxCustomGridTableItem *Sender,
		  TcxCustomGridRecord *ARecord, const Variant &AValue,
          TcxEditValidateInfo *AData)
{
  String AErrorText;
  if (DoPhoneNumberValidate(AValue, AErrorText))
  {
	AData->ErrorType = eetWarning;
	String AFullName = GetPersonFullName(ARecord);
	if (Trim(AFullName) > "")
	{
	  AData->ErrorText = AErrorText + " for " + AFullName;
	}
	else
	{
	  AData->ErrorText = AErrorText;
	}
  }
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::cxHintStyleControllerShowHintEx(TObject *Sender,
		  String &Caption, String &HintStr, bool &CanShow, THintInfo &HintInfo)

{
	stGrid->Header->Glyph->Assign(NULL);
	TObject *AViewInfo = static_cast<TObject*>(HintInfo.HintData);
	if (dynamic_cast<TcxGridDataCellViewInfo*>(AViewInfo))
	{
		switch ((static_cast<TcxGridDataCellViewInfo*>(AViewInfo))->EditViewInfo->ErrorData->ErrorType)
		{
			case eetError: {stGrid->Header->Glyph->Assign(cxEditErrorIcon); break;}
			case eetWarning: {stGrid->Header->Glyph->Assign(cxEditWarningIcon); break; }
			case eetInfo: {stGrid->Header->Glyph->Assign(cxEditInfoIcon); break; }
			case eetCustom: {stGrid->Header->Glyph->Assign(icCustomIcon1->Picture->Bitmap); break; }
		}
	}
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::FormCreate(TObject *Sender)

{
  InitializeEditors(this);
}
//---------------------------------------------------------------------------

void __fastcall TEditorsInPlaceValidationDemoMainForm::InitializeEditors(TObject *Sender)

{
  TcxEditValidationOptions AValidationOptions;
  if (miValidationRaiseException->Checked)
  {
	AValidationOptions << evoRaiseException;
  }
  if (miValidationShowErrorIcon->Checked)
  {
	AValidationOptions << evoShowErrorIcon;
  }
  if (miValidationAllowLoseFocus->Checked)
  {
	AValidationOptions << evoAllowLoseFocus;
  }

  cxGridTableViewColumnFirstName->Properties->ValidationOptions = AValidationOptions;
  cxGridTableViewColumnLastName->Properties->ValidationOptions = AValidationOptions;
  cxGridTableViewColumnAddress->Properties->ValidationOptions = AValidationOptions;
  cxGridTableViewColumnPhoneNumber->Properties->ValidationOptions = AValidationOptions;
  cxGridTableViewColumnEmail->Properties->ValidationOptions = AValidationOptions;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoAddressValidate(const Variant &AValue, TCaption &AErrorText)
{
  int I = (dynamic_cast<TcxComboBoxProperties*> (cxGridTableViewColumnAddress->Properties))->Items->IndexOf(VarToStr(AValue));
  bool rv = (I == -1);
  if (rv)
  {
	AErrorText = "Please select an address from the list";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoEmailValidate(const Variant &AValue, TCaption &AErrorText)
{
  String S = VarToStr(AValue);
  bool rv = ((S != "")&&(!Cxregexpr::IsTextFullValid(S, "[A-z0-9_-]+@[A-z0-9_-]+\\.[A-z0-9_-]+(\\.[A-z]+)*")));
  if (rv)
  {
	AErrorText = "Please enter a valid email address";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoFirstNameValidate(const Variant &AValue, TCaption &AErrorText)
{
  bool rv = (VarToStr(AValue) == "");
  if (rv)
  {
	AErrorText = "Please enter a value";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoLastNameValidate(const Variant &AValue, TCaption &AErrorText)
{
  bool rv = (VarToStr(AValue) == "");
  if (rv)
  {
	AErrorText = "Please enter a value";
  }
  return rv;
}
//---------------------------------------------------------------------------

bool TEditorsInPlaceValidationDemoMainForm::DoPhoneNumberValidate(const Variant &AValue, TCaption &AErrorText)
{
  String S = VarToStr(AValue);
  bool rv = ((S == "")||(!Cxregexpr::IsTextValid(S, "(\\(\\d\\d\\d\\)' ')?\\d\\d\\d-\\d\\d\\d\\d")));
  if (rv)
  {
	AErrorText = "Please enter a valid phone number";
  }
  return rv;
}
//---------------------------------------------------------------------------

String TEditorsInPlaceValidationDemoMainForm::GetPersonFullName(TcxCustomGridRecord *ARecord)
{
  String AFirstName = VarToStr(ARecord->Values[cxGridTableViewColumnFirstName->Index]);
  String ALastName = VarToStr(ARecord->Values[cxGridTableViewColumnLastName->Index]);
  if ((Trim(AFirstName) > "")&&(Trim(ALastName) > ""))
  {
	return AFirstName + " " + ALastName;
  }
  else
  {
	return AFirstName + ALastName;
  }
}
//---------------------------------------------------------------------------


